Modules/grpmodule.c

```diff
@@ -119,16 +121,67 @@ grp_getgrgid_impl(PyObject *module, PyObject *id)
         }
         Py_DECREF(py_int_id);
     }
+#ifdef HAVE_GETGRGID_R
+    Py_BEGIN_ALLOW_THREADS
+    int status;
+    Py_ssize_t bufsize;
+    struct group grp;
+
+    bufsize = sysconf(_SC_GETGR_R_SIZE_MAX);
+    if (bufsize == -1) {
+        bufsize = 1024;
+    }
+    buf = PyMem_RawMalloc(bufsize);
+
+    while (1) {
+        status = getgrgid_r(gid, &grp, buf, bufsize, &p);
+        if (p != NULL || status != ERANGE) {
+            break;
+        }
+        if (bufsize > (PY_SSIZE_T_MAX >> 1)) {
+            PyErr_NoMemory();
+            nomem = 1;
+            status = ERANGE;
```

Comment:
I don’t see the point of calling PyErr_NoMemory when you set RuntimeError later on.

Also, setting _status_ seems redundant because you only get here if you get ERANGE.

Reply:
You're right about ERANGE, will fix.

@serhiy-storchaka you proposed the PyErr_NoMemory, how should I properly raise that error? I used RuntimeError afterwards but that does not seem right.

Thank you

Reply:
Wouldn't MemoryError be more appropriate that RuntimeError?

Reply:
We use MemoryError on integer overflow when allocating memory. We do it all the time in unicodeobject.c for example.